Content.Server/Explosion/EntitySystems/ExplosionSystem.GridMap.cs

```diff
@@ -0,0 +1,374 @@
+namespace Content.Server.Explosion.EntitySystems;
+
+// This partial part of the explosion system has all of the functions used to facilitate explosions moving across grids.
+// A good portion of it is focused around keeping track of what tile-indices on a grid correspond to tiles that border
+// space. AFAIK no other system currently needs to track these "edge-tiles". If they do, this should probably be a
+// property of the grid itself?
+public sealed partial class ExplosionSystem : EntitySystem
+{
+    /// <summary>
+    ///     Set of tiles of each grid that are directly adjacent to space, along with the directions that face space.
+    /// </summary>
+    private Dictionary<GridId, Dictionary<Vector2i, NeighborFlag>> _gridEdges = new();
+
+    /// <summary>
+    ///     On grid startup, prepare a map of grid edges.
+    /// </summary>
+    private void OnGridStartup(GridStartupEvent ev)
+    {
+        if (!_mapManager.TryGetGrid(ev.GridId, out var grid))
+            return;
+
+        Dictionary<Vector2i, NeighborFlag> edges = new();
+        _gridEdges[ev.GridId] = edges;
+
+        foreach (var tileRef in grid.GetAllTiles())
+        {
+            if (tileRef.Tile.IsEmpty)
+                continue;
+
+            if (IsEdge(grid, tileRef.GridIndices, out var dir))
+                edges.Add(tileRef.GridIndices, dir);
+        }
+    }
+
+    private void OnGridRemoved(GridRemovalEvent ev)
+    {
+        _airtightMap.Remove(ev.GridId);
+        _gridEdges.Remove(ev.GridId);
+    }
+
+    /// <summary>
+    ///     Take our map of grid edges, where each is defHined in their own grid's reference frame, and map those
+    ///     edges all onto one grids reference frame.
+    /// </summary>
+    public (Dictionary<Vector2i, BlockedSpaceTile>, float) TransformGridEdges(MapId targetMap, GridId? referenceGrid, List<GridId> localGrids)
+    {
+        Dictionary<Vector2i, BlockedSpaceTile> transformedEdges = new();
+
+        var targetMatrix = Matrix3.Identity;
+        Angle targetAngle = new();
+        var tileSize = (float) DefaultTileSize;
+
+        // if the explosion is centered on some grid (and not just space), get the transforms.
+        if (referenceGrid != null)
+        {
+            var targetGrid = _mapManager.GetGrid(referenceGrid.Value);
+            var xform = Transform(targetGrid.GridEntityId);
+            targetAngle = xform.WorldRotation;
+            targetMatrix = xform.InvWorldMatrix;
+            tileSize = (float) targetGrid.TileSize;
+        }
+
+        var offsetMatrix = Matrix3.Identity;
+        offsetMatrix.R0C2 = tileSize / 2;
+        offsetMatrix.R1C2 = tileSize / 2;
+
+        // Here we can end up with a triple nested for loop:
+        // foreach other grid
+        //   foreach edge tile in that grid
+        //     foreach tile in our grid that touches that tile (vast majority of the time: 1 tile, but could be up to 4)
+
+        HashSet<Vector2i> transformedTiles = new();
```

Comment:
Currently each grid tiles can block/intersecting up to 4 space-tiles, though quite often it will just be 2-3 unique tiles, I just use hash-set to as a lazy way of adding all 4 tiles & automatically removing duplicates. I don't know how to just make this an array w/o adding explicit duplicate-checking code, so this is probably fine?